src/bin/common/solving_loop.ml

```diff
@@ -996,43 +1324,136 @@ let main () =
-          | _ -> st
-        in
-        (* TODO:
-           - Separate statements that should be ignored from unsupported
-             statements and throw exception or print a warning when an
-             unsupported statement is encountered.
-        *)
-        let cnf =
-          D_cnf.make (State.get State.logic_file st).loc
-            (State.get solver_ctx_key st).ctx td
-        in
-        State.set solver_ctx_key (
-          let solver_ctx = State.get solver_ctx_key st in
-          { solver_ctx with ctx = cnf }
-        ) st
+        Printer.print_dbg ~header:true
+          "Ignoring statement: %a" Typer_Pipe.print td;
+        st
   in
-  let handle_stmts all_context st l =
+
+  (* Handle each statement one after the other.
+     Still experimental due to push & pop issues. *)
+  let handle_stmts_full_incremental all_context st l =
     let rec aux named_map st = function
       | [] -> State.set named_terms named_map st
       | stmt :: tl ->
-        let st = handle_stmt all_context st stmt in
+        let st = handle_stmt_full_incremental all_context st stmt in
         let named_map = add_if_named ~acc:named_map stmt in
         aux named_map st tl
     in
     aux (State.get named_terms st) st l
   in
+
+  let get_current_path = state_get ~default:[] current_path_key in
+
+  let get_pushed_paths =
+    state_get ~default:(Vec.create ~dummy:[]) pushed_paths_key
+  in
+
+  (* Same as handle_stmts_full_incremental, but removes pops and pushes.
+     Instead, every command that perform an effect on the environment
+     (definitions, assertion, resets, set option, ...) are stored in a list of
+     commands. When a (push n) is performed, this list is stored n times in a
+     Vec. When a pop m occurs, we reset the analysis, the Vec is popped m times
+     and the list of commands is replayed before continuing the analysis.
+     /!\ Pop do not reset options, so there may be inconstencies between a first
+     execution and a reinitialized one. The handle_reset function should handle
+     this, but it is not the case yet. *)
+  let handle_stmt_pop_reinit all_context st l =
+    let rec aux named_map st = function
+      | [] -> State.set named_terms named_map st
+      | (stmt: stmt) :: tl ->
+        begin
+          let current_path = get_current_path st in
+          match stmt with
+          | {contents = `Push n; _} ->
+            let pushed_paths = get_pushed_paths st in
+            for _ = 0 to n - 1 do
+              Vec.push pushed_paths current_path
+            done;
+            aux named_map st tl
+          | {contents = `Pop n; _} ->
+            let pushed_paths = get_pushed_paths st in
+            let rec pop_until until res =
+              if until = 0 then
+                res
+              else
+                pop_until (until - 1) (Vec.pop pushed_paths)
+            in
+            let prefix = pop_until (n - 1) (Vec.pop pushed_paths) in
+            let st = handle_reset st in
```

Comment:
I don't think it is right to use `handle_reset` here; some things should not be reset (such as the step counter — we did do the previous steps and we should count them, even if we are redoing them).

Reply:
Should `(reset)` reset the Steps counter actually?

Reply:
Yes, the specification says:

> (reset) resets the solver completely to the state it had after it was started and before it started
reading commands.
> This allows the user to reset the state of the solver without paying the cost of restarting it.

Basically the goal is that calling alt-ergo on a script that does `<script1>(reset)<script2>` should behave identically as calling alt-ergo on `<script1>` then a separate instance on `<script2>`.

(It is not the case currently because we don't reset counters and weak tables, but in any case that is a different specification than what is needed for replaying statements in this PR and so the same function should not be used)